sheets/enrollment_change_api.py

```diff
@@ -0,0 +1,266 @@
+        """
+        For all successfully-processed rows, programatically sets the completion date column and
+        blanks the error column.
+
+        Args:
+            row_results (iterable of RowResult): Objects representing the results of processing a row
+        """
+        for row_result in row_results:
+            self.worksheet.update_values(
+                crange="{processor_col}{row_index}:{error_col}{row_index}".format(
+                    processor_col=self.sheet_metadata.PROCESSOR_COL_LETTER,
+                    error_col=self.sheet_metadata.ERROR_COL_LETTER,
+                    row_index=row_result.row_index,
+                ),
+                values=[
+                    [
+                        ENROLL_CHANGE_SHEET_PROCESSOR_NAME,
+                        format_datetime_for_sheet_formula(
+                            row_result.row_db_record.date_completed.astimezone(
+                                settings.SHEETS_DATE_TIMEZONE
+                            )
+                        ),
+                        "",
+                    ]
+                ],
+            )
+
+    def update_row_errors(self, failed_row_results):
+        """
+        For all rows that could not be processed, sets the error column value to an error message.
+
+        Args:
+            failed_row_results (iterable of RowResult): Objects representing the results of processing a row
+        """
+        for row_result in failed_row_results:
+            self.worksheet.update_value(
+                "{}{}".format(
+                    self.sheet_metadata.ERROR_COL_LETTER, row_result.row_index
+                ),
+                row_result.message,
+            )
+
+    def get_or_create_request(self, row_data):
+        """
+        Ensures that an object exists in the database that represents the given enrollment change request, and
+        that it reflects the correct state based on the data in the spreadsheet row.
+
+        Args:
+            row_data (list of str): Raw data from a row in the spreadsheet
+
+        Returns:
+            (Type(EnrollmentChangeRequestModel), bool, bool): A tuple containing an object representing the request,
+                a flag that indicates whether or not it was newly created, and a flag that indicates
+                whether or not it was updated.
+        """
+        form_response_id = int(
+            row_data[self.sheet_metadata.FORM_RESPONSE_ID_COL].strip()
+        )
+        user_input_json = json.dumps(
+            self.sheet_metadata.get_form_input_columns(row_data)
+        )
+        with transaction.atomic():
+            enroll_change_request, created = self.request_model_cls.objects.select_for_update().get_or_create(
+                form_response_id=form_response_id,
+                defaults=dict(raw_data=user_input_json),
+            )
+            raw_data_changed = enroll_change_request.raw_data != user_input_json
```

Comment:
This compares the JSON strings instead of the encoded value, so for example dictionaries may not match each other. Would that be a problem here?

Reply:
The raw row data is always going to be a list of strings, so the ordering of elements will never be an issue